### PR TITLE
[DO NOT MERGE] Support for Search API views

### PR DESCRIPTION
**Pull request by mtodor**
Added support for Search API integration.

Features:
- support for config per field (for example Facets configuration)
- introduce configuration template collection
- added simple check script that checks correct configuration relations (development helper script)

### CODE REVIEW

File: scripts/check_search_api_configs.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+ * Use it with: drush scr
+ * -> Be cause we expect Drupal to be bootstrapped.
+ */
+
+use Drupal\block\Entity\Block;
+use Drupal\facets\Entity\Facet;
+use Drupal\search_api\Entity\Index;
+use Drupal\views\Entity\View;
+
+echo "Checking Search API configurations..." . PHP_EOL . PHP_EOL;
```
Comment (dbosen) [inline scripts/check_search_api_configs.php:16]: Nit: single quotes would be fine, here and some other places in this file, where no variable is expanded.

File: src/ConfigTemplateDefinitionCollection.php [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+   *   The config template collection configuration file.
+   *
+   * @return \Drupal\testsite_builder\ConfigTemplateDefinitionCollection
+   *   Returns instance of config template definition collection.
+   *
+   * @throws \Exception
+   */
+  public static function createFromFile($file_name): ConfigTemplateDefinitionCollection {
+    $instance = static::create(\Drupal::getContainer());
+    $collection = $instance->yamlSerializer->decode(file_get_contents($file_name));
```
Comment (dbosen) [inline src/ConfigTemplateDefinitionCollection.php:114]: Ah, the static decode call again. Here as well as in line 178

File: src/ConfigTemplateDefinitionResolver.php [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+   *   The instance of config template definition resolver.
+   */
+  public static function create(ContainerInterface $container, ConfigTemplateDefinitionCollection $collection): ConfigTemplateDefinitionResolver {
+    $instance = new static(
+      $container->get('testsite_builder.config_template_type_manager'),
+      $container->get('entity_type.manager'),
+      $container->get('entity_field.manager')
+    );
+
+    $instance->collection = $collection;
```
Comment (dbosen) [inline src/ConfigTemplateDefinitionResolver.php:102]: I might be missing something here, but this creation pattern makes only sense, if you either create the $instance from a parent class, as you do in Drupal\testsite_builder\Plugin\testsite_builder\EntityType\Media::create(). Or when you have multiple factories, as in ConfigTemplateDefinitionCollection::createFromTemplates().

In this case, you can just add the $collection to the __construct parameters and handle it there.

File: src/ConfigTemplateDefinitionResolver.php [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+   * @param string $bundle
+   *   The bundle name.
+   * @param \Drupal\Core\Field\FieldDefinitionInterface $field_definition
+   *   The field definition.
+   * @param array $field_templates
+   *   The list of configuration templates for field.
+   *
+   * @throws \Exception
+   */
+  protected function createAllFieldConfigurations(string $entity_type, string $bundle, FieldDefinitionInterface $field_definition, array $field_templates) {
```
Comment (dbosen) [inline src/ConfigTemplateDefinitionResolver.php:274]: $entity_type is unused

File: src/ConfigTemplateDefinitionResolver.php [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+    $dynamic_bundle_definitions = $this->templateDefinition->getDynamicBundleDefinitions();
+
+    $field_config = $this->createConfig($entity_type, $bundle, $dynamic_bundle_definitions, []);
+    $config_parts = explode('.', $this->templateDefinition->getKey('source'));
+
+    // Create ID for config.
+    array_pop($config_parts);
+    $config_id = "{$this->collection->getId()}_{$entity_type}_{$bundle}_{$field_name}";
+    $field_config['id'] = $config_id;
+    array_push($config_parts, $config_id);
```
Comment (dbosen) [inline src/ConfigTemplateDefinitionResolver.php:317]: nit: $config_parts[] = $config_id is faster ;-)

File: src/ConfigTemplateMerge.php [hunk 6]
```diff
@@ -18,6 +18,7 @@ class ConfigTemplateMerge {
   const ADD_VALUE = 1;
   const ADD_KEY = 2;
   const CHANGE_VALUE = 3;
+  const UPDATE_VALUE = 4;
```
Comment (dbosen) [inline src/ConfigTemplateMerge.php:21]: Same as with variables and functions it would be better to provide the public access explicitly.

**Review by dbosen — COMMENTED** [review]
I just quickly added some small things, that occurred to me when just glancing over the code for the first time.
I will need to do a much more in-depth review tomorrow.

File: src/ConfigTemplateDefinitionResolver.php (hunk 3 above)
Comment (ol0lll) [reply src/ConfigTemplateDefinitionResolver.php:102]: fixed in #37

File: scripts/check_search_api_configs.php (hunk 1 above)
Comment (ol0lll) [reply scripts/check_search_api_configs.php:16]: fixed in #38

File: src/ConfigTemplateDefinitionCollection.php (hunk 2 above)
Comment (ol0lll) [reply src/ConfigTemplateDefinitionCollection.php:114]: fixed in #37

File: src/ConfigTemplateDefinitionResolver.php (hunk 4 above)
Comment (ol0lll) [reply src/ConfigTemplateDefinitionResolver.php:274]: fixed in #37

File: src/ConfigTemplateMerge.php (hunk 6 above)
Comment (ol0lll) [reply src/ConfigTemplateMerge.php:21]: fixed in #37

File: src/ConfigTemplateDefinitionResolver.php (hunk 5 above)
Comment (ol0lll) [reply src/ConfigTemplateDefinitionResolver.php:317]: config_parts was removed

**Pull request closed without merging**